beacon-chain/db/kv/state.go

```
@@ -65,21 +63,21 @@ func (s *Store) StateOrError(ctx context.Context, blockRoot [32]byte) (state.Bea
 	return st, nil
 }
 
-// GenesisState returns the genesis state in beacon chain.
 func (s *Store) GenesisState(ctx context.Context) (state.BeaconState, error) {
-	ctx, span := trace.StartSpan(ctx, "BeaconDB.GenesisState")
-	defer span.End()
-
-	cached, err := genesis.State(params.BeaconConfig().ConfigName)
+	st, err := genesis.State()
 	if err != nil {
-		tracing.AnnotateError(span, err)
-		return nil, err
-	}
-	span.SetAttributes(trace.BoolAttribute("cache_hit", cached != nil))
-	if cached != nil {
-		return cached, nil
+		log.WithError(err).Error("Failed to load genesis state")
```

I wonder why we need to log instead of just returning an error here, what's the point of the error field i don't think any part of the function returns the error now

---

This code path could always return `nil, nil` and some places (like execution client and tests) assume it will do so when a genesis state hasn't been explicitly saved, for instance before the chain has started. So this is swallowing an error that won't ever happen in mainnet (embedded genesis) and removes the need for updating a ton of tests. I do plan on revisiting this.

---

updated here: https://github.com/OffchainLabs/prysm/pull/15470/commits/e1630645c91c273f2f452028a1e5e07118375f06

With this change the db will only fall back to the previous state of returning `nil, nil` in test code paths where no explicit initialization of the genesis state has occurred. On testnets and devnets where an unexpected error can occur at runtime (like the previously stored genesis state file can't be read) the db method will return the error.